### PR TITLE
fix: decodeEventLog topics mismatch

### LINKED ISSUE #180: bug: decodeEventLog topics type error when cycling getLogs data
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package Version

^0.1.7

### Current Behavior

> Type 'TopicFilter' is not assignable to type '[signature: `0x${string}`, ...args: LogTopic[]]'.
  Source provides no match for required element at position 0 in target.ts(2322)

### Expected Behavior

No type error

### Steps To Reproduce

```
const logs = await client.getLogs();

logs.forEach((log) => {
  const eventData = decodeEventLog({
    abi: [],
    data: log.data,
    topics: log.topics,
  });
});
```

### Link to Minimal Reproducible Example (StackBlitz, CodeSandbox, GitHub repo etc.)

https://stackblitz.com/edit/viem-getting-started-vqwqeb?file=index.ts

### Anything else?

_No response_

### LINKED ISSUE #197: bug:  decodeEventLog silently fails to decode data when event signature matches but indexed param differs
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package Version

0.1.9

### Current Behavior

the following script should either fails or alternatively be lenient about the log data passed in (it can for example attempt to parse the data field when the topics  length do not matches the indexed arguments)

```js
import { decodeEventLog } from "viem";

const abi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "from",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "to",
        type: "address",
      },
      {
        indexed: true,
        internalType: "uint256",
        name: "id",
        type: "uint256",
      },
    ],
    name: "Transfer",
    type: "event",
  },
];
const parsed = decodeEventLog({
  abi,
  data: "0x0000000000000000000000000000000000000000000000000000000023c34600",
  topics: [
    "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
    "0x000000000000000000000000f39fd6e51aad88f6f4ce6ab8827279cfffb92266",
    "0x00000000000000000000000070e8a65d014918798ba424110d5df658cde1cc58",
  ],
});

console.log(parsed);

```

The script above output the following

```
{
  eventName: 'Transfer',
  args: {
    from: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266',
    to: '0x70E8a65d014918798BA424110d5Df658cdE1cC58'
  }
}
```

See missing third argument ?
`decodeEventLog` expected to parse 3 indexed argument but did not fails when it only parsed 2 of them

Note this event comes form an ERC20 contract who do not have the amount as indexed param
The abi passed in the function is an ERC721 which expect all argument to be indexed.

### Expected Behavior

The function should fails when not all indexed argument are found

Alternatively it could attempt to parse the non-present indexed arguments in the data field

### Steps To Reproduce

```js
import { decodeEventLog } from "viem";

const abi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "from",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "to",
        type: "address",
      },
      {
        indexed: true,
        internalType: "uint256",
        name: "id",
        type: "uint256",
      },
    ],
    name: "Transfer",
    type: "event",
  },
];
const parsed = decodeEventLog({
  abi,
  data: "0x0000000000000000000000000000000000000000000000000000000023c34600",
  topics: [
    "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
    "0x000000000000000000000000f39fd6e51aad88f6f4ce6ab8827279cfffb92266",
    "0x00000000000000000000000070e8a65d014918798ba424110d5df658cde1cc58",
  ],
});

console.log(parsed);

```

### Link to Minimal Reproducible Example (StackBlitz, CodeSandbox, GitHub repo etc.)

https://stackblitz.com/edit/viem-getting-started-oryrtm?file=index.ts

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by wighawag**
Fixes #197
Fixes #180

### CODE REVIEW

**Comment by jxom** [comment]
Thanks for the PR! Just added some refactors I've been meaning to do which will also close #180 as well.

File: src/utils/abi/decodeEventLog.ts [hunk 1]
```diff
@@ -64,14 +66,16 @@ export function decodeEventLog<
   let args: any = isUnnamed ? [] : {}
 
   // Decode topics (indexed args).
-  for (let i = 0; i < inputs.length; i++) {
-    const param = inputs[i]
-    const topic = argTopics[i]
-    if (topic === null) args[param.name || i] = null
```
Comment (jxom) [inline src/utils/abi/decodeEventLog.ts:70]: Removed as a topic on a log cannot be `null`.

File: src/utils/abi/decodeEventLog.ts [hunk 2]
```diff
@@ -64,14 +66,16 @@ export function decodeEventLog<
...
-    if (!topic) continue
-    if (Array.isArray(topic)) {
```
Comment (jxom) [inline src/utils/abi/decodeEventLog.ts:72]: Removed as a topic on a log cannot be an array.

**Comment by wighawag** [comment]
> Thanks for the PR! Just added some refactors I've been meaning to do which will also close #180 as well.

ha yes was wondering about those topic values myself :)
thanks for getting that PR in that quick!

**Pull request merged**